### PR TITLE
Added TransformedRectangularBoundaries class.

### PR DESCRIPTION
**Pull request by MichaelClerx**
See #1451

### CODE REVIEW

File: pints/_transformation.py [hunk 1]
```diff
@@ -940,6 +945,63 @@ def sample(self, n=1):
...
+        b = transformation.to_search(boundaries.upper())
+        lower = np.minimum(a, b)
+        upper = np.maximum(a, b)
+
+        # Pass transformed boundaries to RectangularBoundaries
+        super().__init__(lower, upper)
+
+        # Store input
+        self._bound = boundaries
+        self._trans = transformation
```
Comment (chonlei) [inline pints/_transformation.py:997]: It doesn't matter too much, but maybe use the same naming convention as above? So `self._boundaries` and `self._transform`?

File: pints/tests/shared.py [hunk 2]
```diff
@@ -375,3 +375,23 @@ def n_parameters(self):
         """ See :meth:`pints.Boundaries.n_parameters()`. """
         return self._n_parameters
 
+
+class SwappingTransformation(pints.Transformation):
+    """
+    Transformation that swaps the parameters around (as a very simple example
+    of a transformation that isn't elementwise).
+    """
```
Comment (chonlei) [inline pints/tests/shared.py:383]: Is this useful to have??

**Review by chonlei — APPROVED** [review]
Thanks, Michael. Looks good, but just two minor comments.

File: pints/tests/shared.py (hunk 2 above)
Comment (MichaelClerx) [reply pints/tests/shared.py:383]: It's only for testing! I needed a non-elementwise transformation

File: pints/_transformation.py (hunk 1 above)
Comment (MichaelClerx) [inline pints/_transformation.py:997]:
```suggestion
        self._boundaries = boundaries
        self._transformation = transformation
```

File: pints/_transformation.py [hunk 3]
```diff
@@ -940,6 +945,63 @@ def sample(self, n=1):
...
+
+    def sample(self, n=1):
+        """ See :meth:`Boundaries.sample()`. """
+        # Sample from the original boundaries, but transform to new space
+        return [self._trans.to_search(p) for p in self._bound.sample(n)]
```
Comment (MichaelClerx) [inline pints/_transformation.py:1002]:
```suggestion
        return [self._transformation.to_search(p) 
                for p in self._boundaries.sample(n)]
```

File: pints/_transformation.py [hunk 4]
```diff
@@ -940,6 +945,64 @@ def sample(self, n=1):
...
+        super().__init__(lower, upper)
+
+        # Store input
+        self._boundaries = boundaries
+        self._transformation = transformation
+
+    def sample(self, n=1):
+        """ See :meth:`Boundaries.sample()`. """
+        # Sample from the original boundaries, but transform to new space
+        return [self._transformation.to_search(p)
```
Comment (MichaelClerx) [inline pints/_transformation.py:1002]:
```suggestion
        return [self._transformation.to_search(p)
```

**Pull request merged**